### PR TITLE
Support reading of other NLU formats than Rasa

### LINKED ISSUE #3853: LUIS training data not found
Based on: https://stackoverflow.com/questions/56726192/unable-to-train-luis-model-in-rasa-nlu

**Problem:**
When Rasa is trained it uses the function `rasa/data.py::_is_nlu_file` to check whether a file in the directory is actually a NLU file. The function `_is_nlu_file` currently only works for rasa data, but not for other formats (such das LUIS).

### PR DESCRIPTION
**Pull request by arthurTemporim**
Signed-off-by: ArthurTemporim <arthurrtl@gmail.com>

**Proposed changes**:
- Fix #3853 
- Update Rasa reading files to check the kind of JSON is
- Update the microsoft LUIS exported data from 2.x.x to 3.x.x

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: rasa/data.py [hunk 1]
```diff
@@ -156,15 +155,7 @@ def _is_valid_filetype(path: Text) -> bool:
...
-    with open(file_path, encoding="utf-8") as f:
-        if file_path.endswith(".json"):
-            content = io_utils.read_json_file(file_path)
-            is_nlu_file = (
-                isinstance(content, dict) and content.get("rasa_nlu_data") is not None
-            )
-        else:
-            is_nlu_file = any(_contains_nlu_pattern(l) for l in f)
-    return is_nlu_file
+    return guess_format(file_path) != 'unk'
```
Comment (wochinge) [inline rasa/data.py]: use the defined constant here

File: rasa/data.py [hunk 2]
```diff
@@ -156,15 +155,7 @@ def _is_valid_filetype(path: Text) -> bool:
...
 
 def _is_nlu_file(file_path: Text) -> bool:
-    with open(file_path, encoding="utf-8") as f:
-        if file_path.endswith(".json"):
-            content = io_utils.read_json_file(file_path)
-            is_nlu_file = (
-                isinstance(content, dict) and content.get("rasa_nlu_data") is not None
-            )
-        else:
-            is_nlu_file = any(_contains_nlu_pattern(l) for l in f)
```
Comment (wochinge) [inline rasa/data.py]: you can probably remove the implementation of `_contains_nlu_pattern`, couldn't you?

File: rasa/data.py (hunk 1 above)
Comment (arthurTemporim) [reply rasa/data.py]: I added a constant called `NLU_FORMAT_UNKNOWN` with value "unk" to `constants.py` file because this same value is used in different files

File: rasa/data.py (hunk 2 above)
Comment (arthurTemporim) [reply rasa/data.py]: Sure, now is removed, thanks!

File: rasa/constants.py [hunk 3]
```diff
@@ -12,6 +12,8 @@
 DEFAULT_NLU_RESULTS_PATH = "nlu_comparison_results"
 DEFAULT_REQUEST_TIMEOUT = 60 * 5  # 5 minutes
 
+NLU_FORMAT_UNKNOWN = "unk"
```
Comment (wochinge) [inline rasa/constants.py]: I'd rather reuse the constant from `rasa/nlutraining_data/loading.py` instead of defining a new one.

File: rasa/core/training/interactive.py [hunk 4]
```diff
@@ -797,7 +797,7 @@ def _collect_actions(events: List[Dict[Text, Any]]) -> List[Dict[Text, Any]]:
 
     # need to guess the format of the file before opening it to avoid a read
     # in a write
-    if _guess_format(export_nlu_path) in {"md", "unk"}:
+    if guess_format(export_nlu_path) in {"md", "unk"}:
```
Comment (wochinge) [inline rasa/core/training/interactive.py]: I know this wasn't done by you, but in general we try to not import functions by their name but rather by their module so that it would e.g. be `loading.guess_format`

File: rasa/nlu/training_data/loading.py [hunk 5]
```diff
@@ -28,7 +29,6 @@
 WIT = "wit"
 LUIS = "luis"
 RASA = "rasa_nlu"
-UNK = "unk"
```
Comment (wochinge) [inline rasa/nlu/training_data/loading.py]: I'd actually rather keep the constant here with the others.

File: rasa/constants.py (hunk 3 above)
Comment (wochinge) [reply rasa/constants.py]: ah ok, I'd rather keep it in the old module. if you don't want to import the constant from the other module you could also write a short function `is_nlu_file` to `loading`.

Comment (arthurTemporim) [reply rasa/constants.py]: Done, the solution was to move back `is_nlu_file` to `data.py` and make it  `_is_nlu_flie` import `guess_format` from `loading.py`

File: rasa/core/training/interactive.py (hunk 4 above)
Comment (arthurTemporim) [reply rasa/core/training/interactive.py]: Changed to `loading.guess_format`

File: rasa/nlu/training_data/loading.py (hunk 5 above)
Comment (arthurTemporim) [reply rasa/nlu/training_data/loading.py]: It is back

File: rasa/nlu/__init__.py [hunk 6]
```diff
@@ -2,8 +2,6 @@
 
 import rasa
 from rasa.nlu.train import train
-from rasa.nlu.test import run_evaluation as test
```
Comment (wochinge) [inline rasa/nlu/__init__.py]: why did you remove these?

File: rasa/nlu/training_data/loading.py [hunk 7]
```diff
@@ -123,8 +123,15 @@ def _load(filename: Text, language: Optional[Text] = "en") -> Optional["Training
         return None
 
 
-def _guess_format(filename: Text) -> Text:
-    """Applies heuristics to guess the data format of a file."""
+def guess_format(filename: Text) -> Text:
+    """Applies heuristics to guess the data format of a file.
+
+    Args:
+        filename: Text type with the file name.
```
Comment (wochinge) [inline rasa/nlu/training_data/loading.py]: I think we shouldn't put the name in the docstring since it's already in the code. Having it twice means also double the maintenance effort. I think something like `filename: file whose type should be guessed` would be better

File: rasa/nlu/training_data/loading.py [hunk 8]
```diff
@@ -123,8 +123,15 @@ def _load(filename: Text, language: Optional[Text] = "en") -> Optional["Training
...
+
+    Returns:
+        Text tupe with the guessed nlu format.
```
Comment (wochinge) [inline rasa/nlu/training_data/loading.py]:
```suggestion
        Guessed file format.
```

File: tests/core/test_data.py [hunk 9]
```diff
@@ -105,19 +105,6 @@ def test_same_file_names_get_resolved(tmpdir):
     assert all([f.endswith("stories.md") for f in stories])
```
Comment (wochinge) [inline tests/core/test_data.py]: can you please also add a test which checks whether `_find_core_nlu_files_in_directory` detects luis, wit files?

File: rasa/nlu/training_data/loading.py (hunk 7 above)
Comment (arthurTemporim) [reply rasa/nlu/training_data/loading.py]: Done :D

File: rasa/nlu/__init__.py (hunk 6 above)
Comment (arthurTemporim) [reply rasa/nlu/__init__.py]: If this line is in the code when running `rasa train nlu`, Rasa breaks with this error:
```
ImportError: cannot import name 'get_model' from 'rasa.model' (/home/arthur/Documents/rasa/rasa/rasa/model.py)
```
by removing this line, Rasa works and I didn't find any other issue after this removal

File: tests/core/test_data.py (hunk 9 above)
Comment (arthurTemporim) [reply tests/core/test_data.py]: Yes, the new test is [here](https://github.com/RasaHQ/rasa/blob/39e540c1c15aa6f323d6e722a32285394778ed6e/tests/core/test_data.py#L109)

File: tests/core/test_data.py [hunk 10]
```diff
@@ -105,17 +106,14 @@ def test_same_file_names_get_resolved(tmpdir):
...
-    [
-        "##intent:aintent",
-        "##synonym: synonym",
-        "##regex:a_regex",
-        " ##lookup:additional",
-    ],
-)
-def test_contains_nlu_pattern(line):
-    assert data._contains_nlu_pattern(line)
+def test_find_core_nlu_files_in_directory():
```
Comment (wochinge) [inline tests/core/test_data.py]: Can you have the test case reflect what it's doing? E.g. `test_finding_of_nlu_files_with_multiple_formats`

Comment (wochinge) [reply tests/core/test_data.py]: https://docs.pytest.org/en/latest/parametrize.html

File: rasa/nlu/test.py [hunk 11]
```diff
@@ -19,7 +19,7 @@
 
 from rasa.constants import TEST_DATA_FILE, TRAIN_DATA_FILE
 from rasa import model
-from rasa import train
+from rasa import train as rasa_train
```
Comment (wochinge) [inline rasa/nlu/test.py]: why did you change these lines?

File: tests/core/test_data.py [hunk 12]
```diff
@@ -106,16 +107,48 @@ def test_same_file_names_get_resolved(tmpdir):
...
+        ("wit", {"data/examples/wit/demo-flights.json"}),
     ],
 )
-def test_contains_nlu_pattern(line):
-    assert data._contains_nlu_pattern(line)
+def test_find_core_nlu_files_in_directory(test_input, expected):
+    examples_dir = "data/examples"
+    data_dir = os.path.join(examples_dir, test_input)
+    skill_import = SkillSelector(data_dir)
+    nlu_files = data._find_core_nlu_files_in_directory(data_dir, skill_import)[1]
```
Comment (wochinge) [inline tests/core/test_data.py]: could you also use this function `get_core_nlu_files`, then you don't need to add the SkillSelector instance to the test

File: rasa/data.py [hunk 13]
```diff
@@ -156,21 +154,10 @@ def _is_valid_filetype(path: Text) -> bool:
...
-    return is_nlu_file
-
-
-def _contains_nlu_pattern(text: Text) -> bool:
-    nlu_pattern = r"\s*##\s*(intent|regex||synonym|lookup):"
+    """Checks whether a file is an NLU file."""
+    from rasa.nlu.training_data import loading
```
Comment (wochinge) [inline rasa/data.py]: Can we move that out again and do a top level import, so that it's not imported for every file? To then have working imports you just have to move the import ` from rasa import data` in the `domain` module to `from_directory`

File: CHANGELOG.rst [hunk 14]
```diff
@@ -32,6 +32,7 @@ Fixed
 -----
 - all temporal model files are now deleted after stopping the Rasa server
 - ``rasa shell nlu`` now outputs unicode characters instead of ``\uxxxx`` codes
+- ``rasa train nlu`` now can train different nlu file formats
```
Comment (wochinge) [inline CHANGELOG.rst]: It's not just `rasa train nlu` but rather `rasa train`, so I'd rather go for: 
```suggestion
- ``rasa train`` now also includes NLU files in other formats than the Rasa format
```
